File: java/FiveCardStud.java
import java.util.*;
import java.io.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class FiveCardStud {

	// main method to handle program flow 
	public static void main(String[] args) {
		// initialize a new deck of cards and a list to hold hands
		Deck deck = new Deck();
		List<List<Card>> hands = new ArrayList<>();
		for (int i = 0; i < 6; i++) {
			hands.add(new ArrayList<>()); // create 6 empty hands 
		}
		List<HandRank> handRanks = new ArrayList<>();

		System.out.println("*** P O K E R    H A N D    A N A L Y Z E R ***");	
		System.out.println();
		System.out.println();

		if (args.length > 0) { // part 2 with command line args
			System.out.println("*** USING TEST DECK ***");
            		System.out.println();
            		System.out.println("*** File: " + args[0]);
            		readDeckFromFile(args[0], hands);
		} else { // part 1 w/o command line args 
			System.out.println("*** USING RANDOMIZED DECK OF CARDS ***");
            		System.out.println();
            		System.out.println("*** Shuffled 52 card deck:");
            		deck.shuffle();
            		deck.printDeck(false);
            		deck.dealHands(hands);
            		printSixHands(hands);
            		System.out.println("*** Here is what remains in the deck...");
            		deck.printDeck(true);
            		System.out.println();
		}

		sortHands(hands, handRanks);
        	printWinningHand(hands, handRanks);

	}

	// enumeration for suits 
	public enum Suit {
       		DIAMONDS, CLUBS, HEARTS, SPADES
    	}

	// enumeration for hand ranks 
	public enum HandRank {
        	HIGH_CARD,
        	PAIR,
        	TWO_PAIR,
        	THREE_OF_A_KIND,
        	STRAIGHT,
        	FLUSH,
        	FULL_HOUSE,
        	FOUR_OF_A_KIND,
        	STRAIGHT_FLUSH,
        	ROYAL_FLUSH
    	}

	// convert hand rank enum to string for display
	private static String handRankToString(HandRank rank) {
		switch(rank) {
			case HIGH_CARD: return "High Card";
            		case PAIR: return "Pair";
            		case TWO_PAIR: return "Two Pair";
            		case THREE_OF_A_KIND: return "Three Of A Kind";
            		case STRAIGHT: return "Straight";
            		case FLUSH: return "Flush";
            		case FULL_HOUSE: return "Full House";
            		case FOUR_OF_A_KIND: return "Four Of A Kind";
            		case STRAIGHT_FLUSH: return "Straight Flush";
            		case ROYAL_FLUSH: return "Royal Straight Flush";
            		default: return "Unknown";
		}
	}

	// method to print winning hand 
	private static void printWinningHand(List<List<Card>> hands, List<HandRank> handRanks) {
		System.out.println("--- WINNING HAND ORDER ---");
		for (int i = 0; i < hands.size(); i++) {
			List<Card> hand = hands.get(i);
			for(int j = 0; j < hand.size(); j++) {
				System.out.print(hand.get(j).toString() + " ");
			}
			System.out.println("- " + handRankToString(handRanks.get(i)));
		}
		System.out.println();
	}

	// method to print out the 6 hands dealt 
	private static void printSixHands(List<List<Card>> hands) {
		System.out.println("*** Here are the six hands...");
		for (int i = 0; i < 6; i++) {
			List<Card> hand = hands.get(i);
			for (int j = 0; j < hand.size(); j++) {
				System.out.print(hand.get(j).toString() + " ");
			}
			System.out.println();
		}
		System.out.println();
	}

	// method to clear existing hand ranks and sort hands based on their rank
	private static void sortHands(List<List<Card>> hands, List<HandRank> handRanks) {
		handRanks.clear();
		Collections.sort(hands, new Comparator<List<Card>>() { // used to compare two hands by rank
			public int compare(List<Card> hand1, List<Card> hand2) {
				if (compareHands(hand1, hand2)) return -1;
				else return 1;
			}
		});

		for (List<Card> hand : hands) {
			handRanks.add(classifyHand(hand));
		}
	}

	// method to classify a hand by rank 
	public static HandRank classifyHand(List<Card> hand) {
		// create temp sortedHand to store hand in descending order for easy classification
		List<Card> sortedHand = new ArrayList<>(hand);
		Collections.sort(sortedHand, new Comparator<Card>() { 
			public int compare(Card a, Card b) {
				return Integer.compare(b.getRank(), a.getRank()); // descending order 
			}
		});

		// keep track of suits/ranks occurences 
		int[] occurrences = new int[15];
		int[] suitCount = new int[4];
		char[] suits = {'D', 'C', 'H', 'S'};

		for (int i = 0; i < sortedHand.size(); i++) {
			occurrences[sortedHand.get(i).getRank()]++;
			for (int j = 0; j < 4; j++) {
				if (sortedHand.get(i).getSuit() == suits[j]) {
					suitCount[j]++;
					break;
				}
			}
		}

		boolean flush = false;
		for (int i = 0; i < 4; i++) { // checks if hand is a flush (same suit)
			if (suitCount[i] == 5) {
				flush = true;
				break;
			}
		}
		boolean straight = isStraight(hand);
		
		if (flush && straight) {
			Set<Integer> royalRanks = new HashSet<>(Arrays.asList(14, 13, 12, 11, 10));
			Set<Integer> uniqueRanks = new HashSet<>();
			for (int i = 0; i < sortedHand.size(); i++) {
				uniqueRanks.add(sortedHand.get(i).getRank());
			}

			if (uniqueRanks.equals(royalRanks)) {
				return HandRank.ROYAL_FLUSH;
			} else {
				return HandRank.STRAIGHT_FLUSH;
			}	
		}

		int threeCount = 0;
		int pairCount = 0;
		for (int i = 2; i <= 14; i++) {
			if (occurrences[i] == 4) return HandRank.FOUR_OF_A_KIND;
			if (occurrences[i] == 3) threeCount++;
			if (occurrences[i] == 2) pairCount++;
		}

		if (threeCount == 1 && pairCount == 1) return HandRank.FULL_HOUSE;
		if (flush) return HandRank.FLUSH;
		if (straight) return HandRank.STRAIGHT;
		if (threeCount == 1) return HandRank.THREE_OF_A_KIND;
		if (pairCount == 2) return HandRank.TWO_PAIR;
		if (pairCount == 1) return HandRank.PAIR;

		return HandRank.HIGH_CARD;
	}

	// method to compare two hands by rank + tie breaking rules 
	public static boolean compareHands(List<Card> hand1, List<Card> hand2) {
		HandRank rank1 = classifyHand(hand1);
		HandRank rank2 = classifyHand(hand2);

		if (rank1 != rank2) {
			return rank1.compareTo(rank2) > 0;
		}

		// tie breaking

		// Case 1: Flush categories
		if (rank1 == HandRank.FLUSH || rank1 == HandRank.STRAIGHT_FLUSH || rank1 == HandRank.ROYAL_FLUSH) {
			Card highestCard1 = getHighestCard(hand1, false);
			Card highestCard2 = getHighestCard(hand2, false);

			if (highestCard1.getRank() != highestCard2.getRank()) {
				return highestCard1.getRank() > highestCard2.getRank();
			}
			return highestCard1.getSuit() > highestCard2.getSuit(); // if identical
		}

		// Case 2: Straights
		if (rank1 == HandRank.STRAIGHT) {
			boolean checkAce1 = isAceLowStraight(hand1);
			boolean checkAce2 = isAceLowStraight(hand2);

			Card highestCard1 = getHighestCard(hand1, checkAce1);
			Card highestCard2 = getHighestCard(hand2, checkAce2);

			if (highestCard1.getRank() != highestCard2.getRank()) {
				return highestCard1.getRank() > highestCard2.getRank();
			}
			return highestCard1.getSuit() > highestCard2.getSuit(); // If identical
		}

		// Case 3: Two pair
		if (rank1 == HandRank.TWO_PAIR) {
			int highPair1 = getHighestPairRank(hand1);
			int highPair2 = getHighestPairRank(hand2);

			if (highPair1 != highPair2) {
				return highPair1 > highPair2;
			}

			int lowPair1 = getLowestPairRank(hand1);
			int lowPair2 = getLowestPairRank(hand2);

			if (lowPair1 != lowPair2) {
				return lowPair1 > lowPair2;
			}

			int kicker1 = getKicker(hand1);
			int kicker2 = getKicker(hand2);
			return hand1.get(kicker1).getSuit() > hand2.get(kicker2).getSuit();    
		}

		// Case 4: Pairs
		if (rank1 == HandRank.PAIR) {
			int pairRank1 = getPairRank(hand1);
			int pairRank2 = getPairRank(hand2);
		
			if (pairRank1 != pairRank2) {
				return pairRank1 > pairRank2;
			}

			Card highVal1 = getHighestNonPairCard(hand1);
			Card highVal2 = getHighestNonPairCard(hand2);

			if (highVal1.getRank() != highVal2.getRank()) {
				return highVal1.getRank() > highVal2.getRank();
			}
			return highVal1.getSuit() > highVal2.getSuit();

		}

		// Case 5: High card
		if (rank1 == HandRank.HIGH_CARD) {
			Card highestCard1 = getHighestCard(hand1, false);
			Card highestCard2 = getHighestCard(hand2, false);

			if (highestCard1.getRank() != highestCard2.getRank()) {
				return highestCard1.getRank() > highestCard2.getRank();
			}
			return highestCard1.getSuit() > highestCard2.getSuit();
		}

		// handle general poker rules 
		if (rank1 == HandRank.FOUR_OF_A_KIND) {
			int quadRank1 = getQuadRank(hand1);
			int quadRank2 = getQuadRank(hand2);
			return quadRank1 > quadRank2;
		}

		if (rank1 == HandRank.THREE_OF_A_KIND) {
			int tripletRank1 = getTripletRank(hand1);
			int tripletRank2 = getTripletRank(hand2);
			return tripletRank1 > tripletRank2;
		}

		if (rank1 == HandRank.FULL_HOUSE) {
			int tripletRank1 = getTripletRank(hand1);
			int tripletRank2 = getTripletRank(hand2);

			if (tripletRank1 != tripletRank2) {
				return tripletRank1 > tripletRank2;
			}

			int pairRank1 = getPairRank(hand1);
			int pairRank2 = getPairRank(hand2);
			return pairRank1 > pairRank2;
		}

		return false;
	}

	// method to handle command line args 
	public static void readDeckFromFile(String filename, List<List<Card>> hands) {
		hands.clear(); // clear any potentially existing hands 
		for (int i = 0; i < 6; i++) {
			hands.add(new ArrayList<>());
		}

		// use Buffered Reader to read from file 
		try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
			String line;
			Set<String> seenCards = new HashSet<>();
			List<String> lines = new ArrayList<>();

			while ((line = br.readLine()) != null) {
				lines.add(line);
				System.out.println(line); // Display file line to user
			}

			for (int i = 0; i < 6; i++) { 
				String[] cardStrings = lines.get(i).split(",");
				for (String cardStr : cardStrings) {
					cardStr = cardStr.trim();

					// duplicate card check
					if (seenCards.contains(cardStr)) {
						System.out.println("\n*** ERROR - DUPLICATED CARD FOUND IN DECK ***\n");
						System.out.println("*** DUPLICATE: " + cardStr + " ***");
						System.exit(1);
					}
					seenCards.add(cardStr);
					
					char suit = cardStr.charAt(cardStr.length() - 1);
					int rank = 0;

					// card rank conversion
					if (cardStr.charAt(0) == '1' && cardStr.charAt(1) == '0') {
						rank = 10;
					} else if(cardStr.charAt(0) >= '2' && cardStr.charAt(0) <= '9') {
						rank = cardStr.charAt(0) - '0';
					} else {
						switch (cardStr.charAt(0)) {
							case 'J': rank = 11; break;
							case 'Q': rank = 12; break;
							case 'K': rank = 13; break;
							case 'A': rank = 14; break;
						}
					}
					hands.get(i).add(new Card(rank, suit));
				}
			}
			System.out.println();
			printSixHands(hands);
		} catch (IOException e) { 
			System.err.println("Error reading file: " + e.getMessage());
		}
	}

	// method to check if a hand is a straight 
	public static boolean isStraight(List<Card> hand) {
		Set<Integer> ranks = new HashSet<>();

		for (int i = 0; i < hand.size(); i++) {
			ranks.add(hand.get(i).getRank());
		}

		List<Integer> sortedRanks = new ArrayList<>(ranks);
		Collections.sort(sortedRanks); // ascending order

		// special check for Ace-low straight (A, 2, 3, 4, 5)
		if (sortedRanks.equals(Arrays.asList(2, 3, 4, 5, 14))) {
			return true; 
		}

		// checks if it is in consecutive order 
		for (int i = 0; i < sortedRanks.size() - 1; i++) {
			if (sortedRanks.get(i + 1) != sortedRanks.get(i) + 1) {
				return false;
			}
		}
		return true;
	}

	// check if a hand is an Ace-low straight
	public static boolean isAceLowStraight(List<Card> hand) {
		for (int i = 0; i < hand.size(); i++) {
			if (hand.get(i).getRank() == 14) {
				return true;
			}
		}
    		return false;
	}

	// get highest number card, includes case of Ace-low straight
	public static Card getHighestCard(List<Card> hand, boolean isAceStraight) {
		Card highestCard = hand.get(0);
		for (int i = 1; i < hand.size(); i++) {
			Card currentCard = hand.get(i);
			if (isAceStraight && currentCard.getRank() == 14) {
				if (highestCard.getRank() == 14 || currentCard.getRank() > highestCard.getRank()) {
					highestCard = currentCard;
				}
			} else {
				if (currentCard.getRank() > highestCard.getRank()) {
					highestCard = currentCard;
				}
			}
		}
		return highestCard;
	}

	// count occurrences of a specific rank in a hand
	public static int countOccurrences(List<Card> hand, int rank) {
		int count = 0;
		for (int i = 0; i < hand.size(); i++) {
			if (hand.get(i).getRank() == rank) {
				count++;
			}
		}
		return count;
	}
	
	// get the highest card not part of a pair
	public static Card getHighestNonPairCard(List<Card> hand) {
		List<Card> tempHand = new ArrayList<>();

		for (int i = 0; i < hand.size(); i++) {
			if (countOccurrences(hand, hand.get(i).getRank()) != 2) {
				tempHand.add(hand.get(i));
			}
		}

		Card highestCard = tempHand.get(0);
		for (int i = 1; i < tempHand.size(); i++) {
			if (tempHand.get(i).getRank() > highestCard.getRank()) {
				highestCard = tempHand.get(i);
			}
		}

		return highestCard;
	}
	
	// identify kicker card in a hand (not part of pair)
	public static int getKicker(List<Card> hand) {
		int kickerIndex = -1;
		int maxVal = 0;
		for (int i = 0; i < hand.size(); i++) {
			if (countOccurrences(hand, hand.get(i).getRank()) == 1 && hand.get(i).getRank() > maxVal) {
				maxVal = hand.get(i).getRank();
				kickerIndex = i;
			}
		}
		return kickerIndex;
	}

	// get the rank (value) of the pair in a hand
	public static int getPairRank(List<Card> hand) {
		for (int i = 0; i < hand.size(); i++) {
			if (countOccurrences(hand, hand.get(i).getRank()) == 2) {
				return hand.get(i).getRank();
			}
		}
		return -1;
	}

	// get the highest rank (value) of pairs in a hand
	public static int getHighestPairRank(List<Card> hand) {
		int highestPair = -1;
		for (int i = 0; i < hand.size(); i++) {
			if (countOccurrences(hand, hand.get(i).getRank()) == 2 && hand.get(i).getRank() > highestPair) {
				highestPair = hand.get(i).getRank();
			}
		}
		return highestPair;
	}

	// get the lowest rank (value) of pairs in a hand
	public static int getLowestPairRank(List<Card> hand) {
		int lowestPair = 15;
		for (int i = 0; i < hand.size(); i++) {
			if (countOccurrences(hand, hand.get(i).getRank()) == 2 && hand.get(i).getRank() < lowestPair) {
				lowestPair = hand.get(i).getRank();
			}
		}

		if (lowestPair == 15) {
			return -1;  
		}
		return lowestPair;
	}

	// get the rank of the three-of-a-kind in a hand
	public static int getTripletRank(List<Card> hand) {
		for (int i = 0; i < hand.size(); i++) {
			if (countOccurrences(hand, hand.get(i).getRank()) == 3) {
				return hand.get(i).getRank();
			}	
		}
		return -1;
	}

	// get the rank of the four-of-a-kind in a hand
	public static int getQuadRank(List<Card> hand) {
		for (int i = 0; i < hand.size(); i++) {
			if (countOccurrences(hand, hand.get(i).getRank()) == 4) {
				return hand.get(i).getRank();
			}
		}		
		return -1;
	}

}

File: Fortran/FiveCardStud.f90
program FiveCardStud
        use DeckModule
        use CardModule
        use HandRankings
        implicit none

        type(Deck) :: my_deck
        type(Card), allocatable :: hands(:,:)
        type(HandRank), dimension(6) :: handRanks
        integer :: sortedIndices(6)
        integer :: i, j
        integer :: argc
        character(len=100) :: filename
        logical :: success

        argc = command_argument_count() ! get # of command line args
 
        print *, "*** P O K E R    H A N D    A N A L Y Z E R ***"
        print *
        print *

        if (argc > 0) then ! part 2 w/ command line args 
                print *, "*** USING TEST DECK ***"
                print *
                call get_command_argument(1, filename)
                print *, "*** File: ", trim(filename)
                call readDeckFromFile(trim(filename), hands)
        else ! part 1 w/o command line args 
                print *, "*** USING RANDOMIZED DECK OF CARDS ***"
                print *       
                print *, "*** Shuffled 52 card deck:"

                call my_deck%initialize()
                call my_deck%shuffle_deck()
                call my_deck%print_deck(.false.)
                call deal_hands(my_deck, hands)
                call printSixHands(hands)
                print *, "*** Here is what remains in the deck..."
                call my_deck%print_deck(.true.)
                print *
        endif

        call sortHands(hands, handRanks, sortedIndices)
        call printWinningHand(hands, handRanks, sortedIndices)

        deallocate(hands)

contains
        ! to print 6 hands of 5 cards 
        subroutine printSixHands(hands)
                implicit none
                type(Card), intent(in) :: hands(:, :)
                integer :: i, j
                character(len=100) :: line

                print *, "*** Here are the six hands..."  
                do j = 1, 6 
                        do i = 1, 5  
                                if (i < 5) then
                                        write(*, '(A)', advance='no') trim(card_to_string(hands(i, j))) // ' '
                                else
                                        write(*, '(A)', advance='no') trim(card_to_string(hands(i, j)))
                                end if
                        end do
                        print *  
                end do
                print *              

        end subroutine printSixHands 

        ! to print the winning hand order 
        subroutine printWinningHand(hands, handRanks, sortedIndices)
                type(Card), intent(in) :: hands(:, :)
                type(HandRank), intent(in) :: handRanks(:)
                integer, intent(in) :: sortedIndices(:)
                integer :: i, j, idx

                print *, "--- WINNING HAND ORDER ---"
                do i = 1, 6
                        idx = sortedIndices(i)
                        do j = 1, 5
                                !write(*, '(A)', advance="no") trim(card_to_string(hands(sortedIndices(i), j)))
                                write(*, '(A)', advance="no") trim(card_to_string(hands(j, idx)))
                                write(*, '(A)', advance="no") " " 
                        enddo
                        print *, " - ", trim(handRankToString(handRanks(idx)))
                enddo
                print *
        end subroutine printWinningHand

        ! handle command line args to read deck from file 
        subroutine readDeckFromFile(filename, hands)
                implicit none
                character(len=*), intent(in) :: filename
                type(Card), allocatable :: hands(:,:)
                character(len=100) :: line
                character(len=20) :: cardStr
                character(len=100) :: lines(6)
                integer :: i, j, rank, suitIndex
                integer :: ioStatus
                character(len=1) :: suit
                logical :: cardSeen(13, 4)  ! Ranks 2-14 mapped to indices 1-13
                character(len=1), parameter :: suits(4) = ['D', 'C', 'H', 'S']
                integer :: unit

                allocate(hands(5, 6))
                cardSeen = .false.

                unit = 10
                open(unit=unit, file=filename, status='old', action='read')
                do i = 1, 6
                        read(unit, '(A)', iostat=ioStatus) lines(i) ! read each line from file 
                        if (ioStatus /= 0) then
                                print *, 'Error reading file: ', filename ! error check, should not occur 
                                stop
                        endif
                        print *, trim(lines(i))  ! display file content to user
                end do
                print *
                close(unit)

                do i = 1, 6  
                        line = trim(lines(i))
                        do j = 1, 5 
                                if (j < 5) then
                                        call extractCard(line, cardStr, ',')
                                else
                                        call extractCard(line, cardStr)
                                end if
                                cardStr = adjustl(trim(cardStr))  ! Remove leading/trailing spaces
                                call parseCard(cardStr, rank, suit)
                                suitIndex = suitToIndex(suit)

                                ! duplicate card check 
                                if (cardSeen(rank - 1, suitIndex)) then
                                        print *, '*** ERROR - DUPLICATED CARD FOUND IN DECK ***'
                                        print *
                                        print *
                                        print *, '*** DUPLICATE: ', trim(cardStr), ' ***'
                                        stop
                                end if
                                cardSeen(rank - 1, suitIndex) = .true.
                                hands(j, i)%rank = rank
                                hands(j, i)%suit = suit
                        end do
                end do
                
                call printSixHands(hands)
        
        end subroutine readDeckFromFile

        ! sort the hands based on ranks using compareHands
        subroutine sortHands(hands, handRanks, sortedIndices)
                implicit none
                type(Card),intent(inout) :: hands(:,:)
                type(HandRank), intent(out) :: handRanks(:)
                integer, intent(out) :: sortedIndices(:)
                integer :: i, j, n
                type(HandRank) :: tempRank
                integer :: tempIndex

                n = size(hands, 2)  ! get the number of hands
                do i = 1, n
                        sortedIndices(i) = i
                enddo

                do i = 1, n
                        handRanks(i) = classifyHand(hands(:, i))
                enddo

                do i = 1, n - 1
                        do j = i + 1, n
                                if (.not. compareHands(hands(:, sortedIndices(i)), hands(:, sortedIndices(j)))) then
                                        tempIndex = sortedIndices(i)
                                        sortedIndices(i) = sortedIndices(j)
                                        sortedIndices(j) = tempIndex
                                endif
                        enddo
                enddo

        end subroutine sortHands

        ! function to classify the rank of a hand
        function classifyHand(hand) result(rank)
                implicit none
                type(Card), intent(in) :: hand(:)
                type(HandRank) :: rank
                integer :: occurrences(15) = 0
                integer :: suitCount(4) = 0
                character(len=1), parameter :: suits(4) = ['D', 'C', 'H', 'S']
                integer :: i, j
                logical :: localFlush, localStraight
                integer :: threeCount, pairCount
                type(Card), allocatable :: sortedHand(:)
                character(len=100) :: classifiedHandStr 
                occurrences = 0
                suitCount = 0
                allocate(sortedHand(size(hand)))
                sortedHand = hand

                call sortDescending(sortedHand)
                do i = 1, size(sortedHand)
                        occurrences(sortedHand(i)%rank) = occurrences(sortedHand(i)%rank) + 1 ! count occurences of card value 
                        do j = 1, 4
                                if (sortedHand(i)%suit == suits(j)) then
                                        suitCount(j) = suitCount(j) + 1 ! count occurences of suit 
                                        exit
                                endif
                        enddo
                enddo

                localFlush = any(suitCount == 5)
                localStraight = isStraight(sortedHand)

                if (localFlush .and. localStraight) then
                        if (occurrences(10) > 0 .and. occurrences(11) > 0 .and. occurrences(12) > 0 .and. occurrences(13) > 0 .and. occurrences(14) > 0) then
                                rank%rank_value = ROYAL_FLUSH
                                return
                        else
                                rank%rank_value = STRAIGHT_FLUSH
                                return
                        endif
                endif
                
                threeCount = 0
                pairCount = 0
                do i = 2, 14
                        if (occurrences(i) == 4) then
                                rank%rank_value = FOUR_OF_A_KIND
                                return
                        endif
                        if (occurrences(i) == 3) threeCount = threeCount + 1
                        if (occurrences(i) == 2) pairCount = pairCount + 1
                enddo

                if (threeCount == 1 .and. pairCount == 1) then
                        rank%rank_value = FULL_HOUSE 
                        return
                endif

                if (localFlush) then
                        rank%rank_value = FLUSH
                        return
                endif

                if (localStraight) then
                        rank%rank_value = STRAIGHT
                        return
                endif

                if (threeCount == 1) then
                        rank%rank_value = THREE_OF_A_KIND 
                        return
                endif

                if (pairCount == 2) then
                        rank%rank_value = TWO_PAIR
                        return
                endif
        
                if (pairCount == 1) then
                        rank%rank_value = PAIR
                        return
                endif
        
                rank%rank_value = HIGH_CARD
                
        end function classifyHand

        ! to compare hands by ranks, tie breaking rules 
        function compareHands(hand1, hand2) result(isHand1Better)
                type(Card), intent(in) :: hand1(:)
                type(Card), intent(in) :: hand2(:)
                logical :: isHand1Better
                type(HandRank) :: rank1, rank2
                integer :: highPair1, highPair2
                integer :: lowPair1, lowPair2
                integer :: kicker1, kicker2
                type(Card) :: highestCard1, highestCard2
                logical :: checkAce1, checkAce2
                integer :: pairRank1, pairRank2
                integer :: quadRank1, quadRank2
                integer :: tripletRank1, tripletRank2

                rank1 = classifyHand(hand1)
                rank2 = classifyHand(hand2)

                if (rank1%rank_value /= rank2%rank_value) then
                        isHand1Better = rank1%rank_value > rank2%rank_value
                        return
                endif

                ! tie breaking rules:

                ! Case 1: Flush categories (Straight Flush, Flush, Royal Flush)
                if (rank1%rank_value == FLUSH .or. rank1%rank_value == STRAIGHT_FLUSH .or. rank1%rank_value == ROYAL_FLUSH) then
                        highestCard1 = getHighestCard(hand1, .false.)
                        highestCard2 = getHighestCard(hand2, .false.)
                        if (highestCard1%rank /= highestCard2%rank) then
                                isHand1Better = highestCard1%rank > highestCard2%rank
                                return
                        endif
                        isHand1Better = highestCard1%suit > highestCard2%suit
                        return
                endif

                ! Case 2: Straights but not flushes
                if (rank1%rank_value == STRAIGHT) then
                        checkAce1 = isAceLowStraight(hand1)
                        checkAce2 = isAceLowStraight(hand2)
                        highestCard1 = getHighestCard(hand1, checkAce1)
                        highestCard2 = getHighestCard(hand2, checkAce2)

                        if (highestCard1%rank /= highestCard2%rank) then
                                isHand1Better = highestCard1%rank > highestCard2%rank
                                return
                        endif
                        isHand1Better = highestCard1%suit > highestCard2%suit
                        return
                endif

                ! Case 3: Two pair
                if (rank1%rank_value == TWO_PAIR) then
                        highPair1 = getHighestPairRank(hand1)
                        highPair2 = getHighestPairRank(hand2)
                        if (highPair1 /= highPair2) then
                                isHand1Better = highPair1 > highPair2
                                return
                        endif

                        lowPair1 = getLowestPairRank(hand1)
                        lowPair2 = getLowestPairRank(hand2)
                        if (lowPair1 /= lowPair2) then
                                isHand1Better = lowPair1 > lowPair2
                                return
                        endif

                        kicker1 = getKicker(hand1)
                        kicker2 = getKicker(hand2)
                        isHand1Better = kicker1 > kicker2
                        return
                endif

                ! Case 4: Pairs
                if (rank1%rank_value == PAIR) then
                        pairRank1 = getPairRank(hand1)
                        pairRank2 = getPairRank(hand2)
                        if (pairRank1 /= pairRank2) then
                                isHand1Better = pairRank1 > pairRank2
                                return
                        endif

                        highestCard1 = getHighestNonPairCard(hand1)
                        highestCard2 = getHighestNonPairCard(hand2)
                        if (highestCard1%rank /= highestCard2%rank) then
                                isHand1Better = highestCard1%rank > highestCard2%rank
                                return
                        endif
                        isHand1Better = highestCard1%suit > highestCard2%suit
                        return
                endif

                ! Case 5: High card
                if (rank1%rank_value == HIGH_CARD) then
                        highestCard1 = getHighestCard(hand1, .false.)
                        highestCard2 = getHighestCard(hand2, .false.)
                        if (highestCard1%rank /= highestCard2%rank) then
                                isHand1Better = highestCard1%rank > highestCard2%rank
                                return
                        endif
                        isHand1Better = highestCard1%suit > highestCard2%suit
                        return
                endif

                ! Direct comparisons for Four of a Kind and Three of a Kind
                if (rank1%rank_value == FOUR_OF_A_KIND) then
                        quadRank1 = getQuadRank(hand1)
                        quadRank2 = getQuadRank(hand2)
                        isHand1Better = quadRank1 > quadRank2
                        return
                endif

                if (rank1%rank_value == THREE_OF_A_KIND) then
                        tripletRank1 = getTripletRank(hand1)
                        tripletRank2 = getTripletRank(hand2)
                        isHand1Better = tripletRank1 > tripletRank2
                        return
                endif

                ! Full House comparison
                if (rank1%rank_value == FULL_HOUSE) then
                        tripletRank1 = getTripletRank(hand1)
                        tripletRank2 = getTripletRank(hand2)
                        if (tripletRank1 /= tripletRank2) then
                                isHand1Better = tripletRank1 > tripletRank2
                                return
                        endif

                        pairRank1 = getPairRank(hand1)
                        pairRank2 = getPairRank(hand2)
                        isHand1Better = pairRank1 > pairRank2
                        return
                endif
                
                isHand1Better = .false.

        end function compareHands

        ! sort the hand in descending order 
        subroutine sortDescending(sortedHand)
                type(Card), allocatable :: sortedHand(:)
                integer :: i, j
                type(Card) :: temp

                do i = 1, size(sortedHand)-1
                        do j = i+1, size(sortedHand)
                                if (sortedHand(i)%rank < sortedHand(j)%rank) then
                                        temp = sortedHand(i)
                                        sortedHand(i) = sortedHand(j)
                                        sortedHand(j) = temp
                                endif
                        enddo
                enddo
        end subroutine sortDescending

        ! sort the hand in ascending order 
        subroutine sortAscending(array) ! bubble sort
                integer, intent(inout) :: array(:)
                integer :: i, j, temp

                do i = 1, size(array) - 1
                        do j = 1, size(array) - i
                                if (array(j) > array(j + 1)) then
                                        temp = array(j)
                                        array(j) = array(j + 1)
                                        array(j + 1) = temp
                                endif
                        enddo
                enddo
        end subroutine sortAscending

        ! check if the hand is a straight 
        function isStraight(hand) result(is_straight)
                type(Card), intent(in) :: hand(:)
                logical :: is_straight
                integer :: ranks(size(hand))
                integer :: i

                do i = 1, size(hand)
                        ranks(i) = hand(i)%rank
                enddo

                call sortAscending(ranks)

                ! considers Ace Low Straight case 
                if (all(ranks == [2, 3, 4, 5, 14])) then
                        is_straight = .true.
                return
                endif

                is_straight = .true.
                do i = 1, size(ranks) - 1
                        if (ranks(i + 1) /= ranks(i) + 1) then
                                is_straight = .false.
                                return
                        endif
                enddo
        end function isStraight 

        
end program FiveCardStud
